### DISCUSSION TITLE
Build fails after upgrading to Next.js 14 (webpack issues) - Works fine in dev mode

### DISCUSSION THREAD

**Discussion by MartinXPN** [vercel/next.js#57535] [App Router]
Seems like there is an issue with the latest version of Next.js. It fails to build with a webpack error. 
Seems like this might be an issue with `firebase-admin` and Next.js working together.

Here is the full error log:
```
   Creating an optimized production build ...
 ⚠ Custom webpack configuration is detected. When using a custom webpack configuration, the Webpack build worker is disabled by default. To force enable it, set the "experimental.webpackBuildWorker" option to "true". Read more: https://nextjs.org/docs/messages/webpack-build-worker-opt-out
Failed to compile.
../app/[lang]/[courseId]/[exerciseId]/page.js from Terser
  x await isn't allowed in non-async function
      ,-[1404:1]
 1404 | /* harmony import */ var firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(3263);
 1405 | /* harmony import */ var firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(72929);
 1406 | var __webpack_async_dependencies__ = __webpack_handle_async_dependencies__([firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__, firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__]);
 1407 | ([firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__, firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__] = __webpack_async_dependencies__.then ? (await __webpack_async_dependencies__)() : __webpack_async_dependencies__);
      :                                                                                                                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 1408 | 
 1409 | 
 1410 | // TODO: find a way to share the DB between firebase-functions / server / client
      `----
  x await isn't allowed in non-async function
      ,-[1523:1]
 1523 | /* harmony import */ var notion_client__WEBPACK_IMPORTED_MODULE_3__ = __webpack_require__(67299);
 1524 | /* harmony import */ var _db__WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(38474);
 1525 | var __webpack_async_dependencies__ = __webpack_handle_async_dependencies__([firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__]);
 1526 | firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__ = (__webpack_async_dependencies__.then ? (await __webpack_async_dependencies__)() : __webpack_async_dependencies__)[0];
      :                                                                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 1527 | 
 1528 | 
      `----
Caused by:
    0: failed to parse input file
    1: Syntax Error
Error: 
  x await isn't allowed in non-async function
      ,-[1404:1]
 1404 | /* harmony import */ var firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(3263);
 1405 | /* harmony import */ var firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(72929);
 1406 | var __webpack_async_dependencies__ = __webpack_handle_async_dependencies__([firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__, firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__]);
 1407 | ([firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__, firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__] = __webpack_async_dependencies__.then ? (await __webpack_async_dependencies__)() : __webpack_async_dependencies__);
      :                                                                                                                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 1408 | 
 1409 | 
 1410 | // TODO: find a way to share the DB between firebase-functions / server / client
      `----
  x await isn't allowed in non-async function
      ,-[1523:1]
 1523 | /* harmony import */ var notion_client__WEBPACK_IMPORTED_MODULE_3__ = __webpack_require__(67299);
 1524 | /* harmony import */ var _db__WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(38474);
 1525 | var __webpack_async_dependencies__ = __webpack_handle_async_dependencies__([firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__]);
 1526 | firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__ = (__webpack_async_dependencies__.then ? (await __webpack_async_dependencies__)() : __webpack_async_dependencies__)[0];
      :                                                                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 1527 | 
 1528 | 
      `----
Caused by:
    0: failed to parse input file
    1: Syntax Error
> Build failed because of webpack errors
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Command "yarn run build" exited with 1
```

Is there any way to fix this?

**Comment by vignesh-gupta** [comment]
Hey @MartinXPN  could add some piece of code causing this issue?

**Reply by MartinXPN** [reply]
Hey @vignesh-gupta 
I was trying to reproduce this behavior on a smaller project. Seems like my initial assumption of this being related to `firebase-admin` is not correct. There is probably something else wrong with the setup.
Do you know what would be the best way of debugging this to find out the cause?
Otherwise, reproducing this becomes almost impossible.

I tried running `next build --debug` but got the same error with no additional info.

**Reply by mikeborozdin** [reply]
@vignesh-gupta

I'm also having this issue and here's the minimal repo for reproducing the problem: https://github.com/mikeborozdin/next14-firebase-admin-issue

Just run:

```bash
npm i
npm run build
```

All worked fine in Next 13 with the experimental server actions.

**Reply by MaksVibe** [reply]
experimental: {
    webpackBuildWorker: true,
  },

**Comment by JulianJorgensen** [comment]
Happens to me too. I also use Firebase and Firebase admin packages. Have been using latest Next 13 in production since May and it's been working great.
I just upgraded to v14 yesterday and saw same issues on prod build.

**Comment by icyJoseph** [comment]
Could you try to turn of swcMinify and see if the issue persists?

**Reply by JulianJorgensen** [reply]
That changed the errors a bit to following:

> ../app/[locale]/(auth)/admin/config/page.js from Terser
Unexpected token name «__webpack_async_dependencies__», expected punc «,» [../app/[locale]/(auth)/admin/config/page.js:1064,97]
    at js_error (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:32780)
    at croak (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:43882)
    at token_error (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:43989)
    at expect_token (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:44168)
    at expect (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:44298)
    at params_or_seq_ (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:55345)
    at expr_atom (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:60866)
    at maybe_unary (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:71619)
    at expr_ops (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:72615)
    at maybe_conditional (/Users/julianjorgensen/sites/selftime.io/node_modules/next/dist/compiled/terser/bundle.min.js:1:72697)

**Reply by JulianJorgensen** [reply]
My next.config.js file looks like this:
```js
const withNextIntl = require('next-intl/plugin')(
  // This is the default (also the `src` folder is supported out of the box)
  './i18n.js'
);

module.exports = withNextIntl({
  pageExtensions: ["tsx", "js"],
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
    config.module.rules.push(
      ...[
        {
          test: /\.yml$/,
          type: "json",
          use: "yaml-loader",
        },
        {
          test: /\.svg$/,
          use: "@svgr/webpack",
        },
      ]
    );
    return config;
  },
  experimental: {
    serverActions: true,
  },
  swcMinify: false
});

// Next bundle analyzer
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})
module.exports = withBundleAnalyzer(module.exports, {})
```

**Reply by icyJoseph** [reply]
Could you try to remove either of the rules you are pushing?

**Reply by MartinXPN** [reply]
I don't think it's caused by the rules. I've posted my `next.config.js` below. 
I get the same error but only have one rule.

**Reply by karthikjn01** [reply]
Hey, has there been any updates on this? I'm running into the same sort of issues, specifically with Prisma & their serverless drivers. The issue I've currently got:

```
...[articleId]/page.js from Terser
Unexpected token name «Promise», expected punc «,» [app/dashboard/...]
```

**Comment by MartinXPN** [comment]
I tried reproducing this on a smaller project but with no luck so far. I've tried integrating `firebase-admin` as well as some ESM libraries. Apparently, using ESM libraries prevents Turbopack from working (but that's a different issue).

Below are the details of the crashing project (the one that is not possible to build).

Here is my `package.json`:
```json
{
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@emotion/cache": "^11.11.0",
    "@emotion/css": "^11.11.0",
    "@emotion/react": "^11.11.0",
    "@emotion/server": "^11.11.0",
    "@emotion/styled": "^11.11.0",
    "@hookform/resolvers": "^3.3.2",
    "@matejmazur/react-katex": "^3.1.3",
    "@mui/icons-material": "^5.14.15",
    "@mui/material": "^5.14.15",
    "@mui/x-date-pickers": "^6.16.3",
    "@mui/x-tree-view": "^6.0.0-beta.0",
    "@sentry/nextjs": "^7.75.1",
    "@svgr/webpack": "^8.1.0",
    "ace-builds": "^1.31.0",
    "algoliasearch": "^4.20.0",
    "axios": "^1.6.0",
    "clipboard-copy": "^4.0.1",
    "cookies-next": "^4.0.0",
    "encoding": "^0.1.13",
    "endent": "^2.1.0",
    "firebase": "^10.5.2",
    "firebase-admin": "^11.11.0",
    "firebaseui": "^6.1.0",
    "katex": "^0.16.9",
    "moment": "^2.29.4",
    "next": "^14.0.0",
    "next-firebase-auth-edge": "^0.9.0",
    "notion-client": "^6.16.0",
    "notion-types": "^6.16.0",
    "prismjs": "^1.29.0",
    "react": "^18.2.0",
    "react-ace": "^10.1.0",
    "react-calendar-heatmap": "^1.9.0",
    "react-countdown": "^2.3.5",
    "react-dom": "^18.2.0",
    "react-drag-drop-files": "^2.3.10",
    "react-hook-form": "^7.47.0",
    "react-hotkeys-hook": "^4.4.1",
    "react-intersection-observer": "^9.5.2",
    "react-markdown": "^9.0.0",
    "react-multi-split-pane": "^0.3.3",
    "react-notion-x": "^6.16.0",
    "react-slick": "^0.29.0",
    "react-spring": "^9.7.3",
    "recharts": "^2.9.0",
    "sharp": "^0.32.6",
    "slick-carousel": "^1.8.1",
    "superjson": "^2.2.0",
    "typescript": "^5.2.2",
    "use-async-effect": "^2.2.7",
    "zod": "^3.22.4"
  },
  "devDependencies": {
    "@types/node": "^20.8.9",
    "@types/prismjs": "^1.26.2",
    "@types/react": "^18.2.33",
    "@types/react-calendar-heatmap": "^1.6.5",
    "@types/react-dom": "^18.2.14",
    "@types/react-slick": "^0.23.11",
    "eslint": "^8.52.0",
    "eslint-config-next": "^14.0.0"
  },
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
  }
}
```

Here is my `tsconfig.json`:
```json
{
  "compilerOptions": {
    "target": "es5",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "allowSyntheticDefaultImports": true,
    "noFallthroughCasesInSwitch": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true,
    "plugins": [{"name": "next"}],
  },
  "include": [
    "next-env.d.ts", ".next/types/**/*.ts", "middleware.ts",
    "src/**/*", "app/**/*", "@types/**/*", "server/**/*"
  ],
  "exclude": ["node_modules", "**/*.spec.ts", "lib"]
}
```

And here is my `next.config.js`:
```js
const {withSentryConfig} = require('@sentry/nextjs');

module.exports = withSentryConfig({
    reactStrictMode: true,
    webpack(config) {
        config.module.rules.push({
            test: /\.svg$/,
            use: ["@svgr/webpack"]
        });

        return config;
    },
    images: {
        remotePatterns: [{
            protocol: 'https',
            hostname: '**.notion.so',
        }, {
            protocol: 'https',
            hostname: '**.amazonaws.com',
        }, {
            protocol: 'https',
            hostname: '**.googleapis.com',
        }],
        domains: ['flagcdn.com'],
        formats: ['image/avif', 'image/webp'],
        minimumCacheTTL: 60 * 60, // cache for 1h
    }
},

// Injected content via Sentry wizard below
{
    // For all available options, see:
    // https://github.com/getsentry/sentry-webpack-plugin#options

    // Suppresses source map uploading logs during build
    silent: true,
    org: process.env.SENTRY_ORG,
    project: process.env.SENTRY_PROJECT,
},
{
    // For all available options, see:
    // https://docs.sentry.io/platforms/javascript/guides/nextjs/manual-setup/

    // Upload a larger set of source maps for prettier stack traces (increases build time)
    widenClientFileUpload: true,

    // Transpiles SDK to be compatible with IE11 (increases bundle size)
    transpileClientSDK: true,

    // Routes browser requests to Sentry through a Next.js rewrite to circumvent ad-blockers (increases server load)
    tunnelRoute: '/monitoring',

    // Hides source maps from generated client bundles
    hideSourceMaps: true,

    // Automatically tree-shake Sentry logger statements to reduce bundle size
    disableLogger: true,
});
```

**Reply by icyJoseph** [reply]
Well, I'd say at this point, try to remove routes until you get a successful build, and then re-add them until you find one with the actual issue :/ not ideal, but given the complexity of the dependencies in NPM projects, this is probably the fastest way...

**Reply by mikeborozdin** [reply]
@icyJoseph I'm also having this problem and this is the minimal codebase that shows the problem - https://github.com/mikeborozdin/next14-firebase-admin-issue

Just run:

```bash
npm i
npm run build
```

**Reply by MartinXPN** [reply]
Seems like I found the root cause but not sure how to fix it yet.
I was using a server action inside a `startTransition` like this:

```tsx
    useEffect(() => {
        if (!id)
            return;

        return startTransition(async () => {
            const data = await fetchData(id);
            console.log('fetched data:', data);
            setData(data);
        });
    }, [id]);
```

`fetchData` was defined in a different `action.ts` file with `'use server';` at the top of the file.

After removing this `useEffect`, the whole error goes away.

**Reply by icyJoseph** [reply]
Right, transitions should not have async work within them. 

- https://react.dev/reference/react/useTransition#react-doesnt-treat-my-state-update-as-a-transition

Await the promise, and then call the transition.

<img width="1170" alt="Screenshot 2023-10-27 at 18 57 46" src="https://github.com/vercel/next.js/assets/21013447/150cb00f-9892-4c22-a847-6e9aa1aefd60">

**Reply by MartinXPN** [reply]
I tried to remove the `startTransition` completely but got the same error.
What's the right way of calling a server action inside a `useEffect`? Seems like that's the most important part that I'm missing.

**Reply by EmmanuelNwafuru** [reply]
You'll always get an error... What I do is, I maintain the default server state of my page.tsx. This allows me to fetch all necessary data I need using async/await. Then client actions goes into the different components that needs it. In other words, my page.tsx are all server compatible, you won't find any state call, eventually listener, and the likes there.

**Comment by dsfpablo** [comment]
HI, I had same problem and just fix it reinstalling some npm packages. In my case I uninstall next14 and clerk (pnpm uninstall next @clerk/nextjs) and reinstall them (pnpm i next @clerk/nextjs). That did the trick for me. Hope it helps.

**Reply by JulianJorgensen** [reply]
thanks for the suggestion, but after using `rm -rf node_modules` and then reinstalling them, the same error still persists.

I will try removing routes as @icyJoseph suggests to isolate where the issue is coming from.

**Reply by dsfpablo** [reply]
I removed node_modules and the error was still there. I did uninstall and reinstall, then it got fixed.

**Comment by bfowle** [comment]
I'm having this same issue. (EDIT: using `firebase-admin` and `firebase` SDK functions in Server Actions; having to call `initializeApp` each time one needs to prepare a FB function seems to be where the issue is thrown, however the stacktrack is very unhelpful)
I did a bisect when it was last properly working: v13.5.6
It broke at v13.5.6-canary.8

**Reply by icyJoseph** [reply]
Sweet! Now we just have to look at the difference between that and the previous canary. I'm out of the office, could you generate one from the released tab and post it here?

**Reply by bfowle** [reply]
Apologies if I'm misunderstanding the ask, but here's the diff between v13.5.6-canary.7 and .8: https://github.com/vercel/next.js/compare/v13.5.6-canary.7...v13.5.6-canary.8

**Comment by mikeborozdin** [comment]
I'm also having this issue and here's the minimal repo for reproducing the problem: https://github.com/mikeborozdin/next14-firebase-admin-issue

Just run:

```bash
npm i
npm run build
```

All worked fine in Next 13 with the experimental server actions.

**Comment by jengkhaw95** [comment]
I faced the same issue and it seems like the problem is direct importing the server action in a `client component`.
I'm not sure if this behavior is intended, so I think sharing the example code here helps.

I'm using firebase-admin to validate user in a server action:

```javascript
"use server";

import config from "@/config"
import admin from "firebase-admin";
import {initializeApp} from "firebase-admin/app";
import {getAuth} from "firebase-admin/auth";

export async myServerAction(payload) {
    const app = initializeApp({
      credential: admin.credential.cert({
        projectId: config.project_id,
        clientEmail: config.client_email,
        privateKey: config.private_key,
      }),
    });
    const firebaseAuth = getAuth(app);
    
    // the rest of the logic
}
``` 

Doesn't work:

```javascript
"use client";

import {myServerAction} from "@/server-actions"

export function MyForm() {
    const handleSubmit=async()=>{
        // handle loading state, etc...
        const payload = {};
        const response = await myServerAction(payload);
        // handles response
    }
    return <form>
        {/* input fields */}
        <button type="button" onClick={()=>handleSubmit()}>Submit</button>
    </form>
}
``` 

I tried to pass the server action into client component via props (instead of directly importing it) and it worked.

My solution:

```javascript
"use client";

export function MyForm({myServerAction}) {
    const handleSubmit=async()=>{
        // handle loading state, etc...
        const payload = {};
        const response = await myServerAction(payload);
        // handles response
    }
    return <form>
        {/* input fields */}
        <button type="button" onClick={()=>handleSubmit()}>Submit</button>
    </form>
}
``` 

I'm not sure what I'm doing is correct, but it works for now.
 `next 14.0.0`
 `firebase-admin 11.11.0`

**Reply by JulianJorgensen** [reply]
Fantastic! This worked for me too! Thanks a bunch @jengkhaw95 !

**Reply by MartinXPN** [reply]
This worked for me as well. Thank you so much!
Not sure if that should be the intended behavior though.

**Reply by bfowle** [reply]
Considering it's been in the docs for a while to be imported into client components (https://nextjs.org/docs/app/api-reference/functions/server-actions#with-client-components) it does not seem like intended behavior.

**Reply by Connor-8** [reply]
This work for me as well. I noticed that only server actions that imported firestore-admin and that were used in a component with 'use client' caused an error.

**Reply by samezzz** [reply]
This worked for me as well. Thank youuuu

**Comment by thewbuk** [comment]
Same problem here. Does anybody have a better solution than to go over every async function painstakingly? 

`5556.js from Terser
  x await isn't allowed in non-async function
        ,-[228540:1]
 228540 | /* harmony import */ var firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(3263);
 228541 | /* harmony import */ var firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(72929);
 228542 | var __webpack_async_dependencies__ = __webpack_handle_async_dependencies__([firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__, firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__]); 
 228543 | ([firebase_admin_app__WEBPACK_IMPORTED_MODULE_0__, firebase_admin_firestore__WEBPACK_IMPORTED_MODULE_1__] = __webpack_async_dependencies__.then ? (await __webpack_async_dependencies__)() : __webpack_async_dependencies__);
        :                                                                                       
                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 228544 | /**
 228545 |  * <div style={{display: "flex", justifyContent: "space-between", alignItems: "center", padding: 16}}>
 228546 |  * <span>`

**Comment by MaksVibe** [comment]
What?

**Reply by thewbuk** [reply]
Running build complains about `Build failed because of webpack errors`. Unfortunately it does not give where the  problem originates from just that it is caused by Terser

**Comment by akwirick** [comment]
I've run into this same issue with firebase-admin and have had to work around it a few times now.   I also found moving the action to a nearby location can also seem to fix the webpack issue. 

Same directory, relative import seems fine
```
import { myAction } from './actions'
```

Different parent, path import is no bueno
```
import { myAction } from '@/app/other/actions'
```

Don't have time to really dig into the why here but it's another data point.

**Reply by baunov** [reply]
Doesn't seem to fix for me

**Comment by Beetroot11** [comment]
Hey everyone, 

Been dealing with the same issue. I had this issue with `"next": 14.0.1` but once I'd updated to `"next": 14.0.3` everything was then working as expected. 

I also made sure that my `next.config.js` has the following for webpack:
` webpack: (config, { isServer }) => { config.resolve.alias['@'] = path.resolve(__dirname, '.'); return config; },`

Do yourself a favour and try this before going to change all imports (as per the comment from @akwirick - value your development time!).

**Comment by KeisukeNagakawa** [comment]
I solved this problem with `firebase-admin` packages.

For me, when server action includes `FieldValue.serverTimestamp` from `friebase-admin/firestore`, it causes error. 

When I removed it by replaced `new Date()`, it solved.

**Reply by thewbuk** [reply]
That seemed to solve it for me as well

**Comment by mozisan** [comment]
As a workaround, dynamic import solved the issue in my case.

```ts
"use server";

import { ... } from "firebase-admin/app";

export function doSomething() {
  ...
};
```

↓

```ts
"use server";

export async function doSomething() {
  const { ... } = await import("firebase-admin/app");
  ...
};
```

**Comment by nonoakij** [comment]
I don't know if this will help, but in my case, I was able to Build by setting `experimental.esmExternals` to `false` in next.config.js.
My current next.config.js looks like this :
```ts
const path = require("path");

/** @type {import('next').NextConfig} */
const nextConfig = {
  output: "standalone",
  reactStrictMode: true,
  experimental: {
    outputFileTracingRoot: path.join(__dirname, "../../"),
    esmExternals: false,
  },
};

module.exports = nextConfig;
```

**Reply by yannxaver** [reply]
This worked for me as well. Do you have a pointer to what could be the unintended side effects of setting this to false? I couldn't find much information about this flag.

**Reply by nonoakij** [reply]
> This standard pushes the web ecosystem forward by enabling smaller package sizes and JavaScript bundles, ultimately leading to a better user experience.
https://nextjs.org/blog/next-12#es-modules-support-and-url-imports

As I read this, maybe the bundle size will be larger. However, I am not sure about those other points either.

ref: https://github.com/vercel/next.js/discussions/27876

**Reply by Jings** [reply]
That did the trick for me as well even though it does not feel like it is the right way.
Hopefully there is something to come when v14 gets stable.

**Comment by kennyAnyi9** [comment]
Hello my seniors, pls could anyone help me resolve this issue?

I built a simple portfolio website and tried deploying on vercel but I had these logs :


```
Creating an optimized production build ...
--
12:05:23.946 | Failed to compile.
12:05:23.947 |  
12:05:23.948 | ./node_modules/html-to-text/lib/html-to-text.mjs + 24 modules
12:05:23.948 | Cannot get final name for export 'encodeXML' of ./node_modules/entities/lib/esm/index.js
12:05:23.948 |  
12:05:23.948 |  
12:05:23.948 | > Build failed because of webpack errors
12:05:23.994 | error Command failed with exit code 1.
12:05:23.995 | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
12:05:24.021 | Error: Command "yarn run build" exited with 1
```

**Reply by abusada** [reply]
Getting the same issue but this is how it looks:

```
./components/apis/api.base.ts + 65 modules
Cannot get final name for export 'default.create' of ../../node_modules/.pnpm/axios@1.6.2/node_modules/axios/index.js
```

it happens when i include a component that is calling server actions from inside a `useTransition` .. still digging into it!
```
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:28:12 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T8103
Binaries:
  Node: 18.18.2
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: 8.11.0
Relevant Packages:
  next: 14.0.4-canary.41
  eslint-config-next: 14.0.3
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.2
Next.js Config:
  output: N/A
  ```

**Comment by MilanMasta** [comment]
I digged a little bit and found that adding `esmExternals: false` to next.config.js will solve the "__webpack_async_dependencies__" problem with server action in next.js version 14

next.config.js file ->

```
/** @type {import('next').NextConfig} */
const nextConfig = {
    experimental: {
        esmExternals: false,
    }
}

module.exports = nextConfig
```

**Comment by cobbvanth** [comment]
The solution to anyone coming across this later is that you are probably accidentally importing a server action in a client component. (Most likely you are incrementally switching to the app directory from the pages directory. 

For me this was caused by including a server action in my `NavBar` component. What I didn't realize was that a lot of now unused pages in my pages directory were importing the navbar. Removing the navbar from the old pages directory pages will solve this.

**Comment by IliyaZafirov** [comment]
I just uninstalled symbol-observable from webpack and all works fine now.

`npm uninstall symbol-observable`

**Comment by grep-o** [comment]
If you have **Dockerfile**, the issue is this `npm ci`. You should use `npm i`
